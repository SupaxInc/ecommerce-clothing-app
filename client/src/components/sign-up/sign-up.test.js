import React from "react";
import { mount, shallow } from "enzyme";

import * as redux from 'react-redux';

import FormInput from "../form-input/form-input";
import SignUp from "./sign-up";

describe('Sign Up component', () => {
    let wrapper;

    // Mock use dispatch
    const useDispatchMock = jest.spyOn(redux, 'useDispatch');
    const dispatch = jest.fn();

    beforeEach(() => {
        useDispatchMock.mockImplementation(() => dispatch);
        wrapper = shallow(<SignUp />);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render sign-up component', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should submit form on click and dispatch the correct action and parameters', () => {
        expect.assertions(1);

        wrapper.find(FormInput).at(0).dive().find('.form-input').simulate('change', {
            target: {
                name: 'displayName',
                value: 'foo'
            }
        });

        wrapper.find(FormInput).at(1).dive().find('.form-input').simulate('change', {
            target: {
                name: 'email',
                value: 'foo@gmail.com'
            }
        });

        wrapper.find(FormInput).at(2).dive().find('.form-input').simulate('change', {
            target: {
                name: 'password',
                value: 'test123'
            }
        });

        wrapper.find(FormInput).at(3).dive().find('.form-input').simulate('change', {
            target: {
                name: 'confirmPassword',
                value: 'test123'
            }
        });

        wrapper.find('form').simulate(
            'submit', {
                preventDefault() {}
            }
        );

        expect(dispatch).toHaveBeenCalledWith({
            type: 'SIGN_UP_START',
            payload: {
                displayName: 'foo',
                email: 'foo@gmail.com',
                password: 'test123'
            }
        });
    });

    it('should submit form on click but do not dispatch the action due to password and confirm password does not equal each other', () => {
        expect.assertions(1);

        wrapper.find(FormInput).at(0).dive().find('.form-input').simulate('change', {
            target: {
                name: 'displayName',
                value: 'foo'
            }
        });

        wrapper.find(FormInput).at(1).dive().find('.form-input').simulate('change', {
            target: {
                name: 'email',
                value: 'foo@gmail.com'
            }
        });

        wrapper.find(FormInput).at(2).dive().find('.form-input').simulate('change', {
            target: {
                name: 'password',
                value: 'test123'
            }
        });

        wrapper.find(FormInput).at(3).dive().find('.form-input').simulate('change', {
            target: {
                name: 'confirmPassword',
                value: 'test1234'
            }
        });

        wrapper.find('form').simulate(
            'submit', {
                preventDefault() {}
            }
        );

        // did NOT call the sign_up_start action
        expect(dispatch).not.toHaveBeenCalledWith({
            type: 'SIGN_UP_START',
            payload: {
                displayName: 'foo',
                email: 'foo@gmail.com',
                password: 'test123'
            }
        });
    });
}); 