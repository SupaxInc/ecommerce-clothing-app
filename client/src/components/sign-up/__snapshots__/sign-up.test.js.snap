// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign Up component should render sign-up component 1`] = `ShallowWrapper {}`;
